compiler-core/src/language_server/server.rs

```diff
@@ -398,6 +411,87 @@ where
             _ = self.changed_projects.insert(project_path);
         }
     }
+
+    fn validate_module_filename(&self, path: &Utf8Path, src: &str) -> Option<Feedback> {
+        // Only validate .gleam files
+        if path.extension().map(|e| e != "gleam").unwrap_or(true) {
+            return None;
+        }
+
+        // Only validate inside a Gleam project. Outside projects the LS may only provide formatting.
+        let project_root = self.router.project_path(path)?;
+
+        let rel = match path.strip_prefix(&project_root) {
+            Ok(p) => p,
+            Err(_) => return None,
+        };
+        let rel_str = rel.as_str();
+        // Trim the leading project area (src/test/dev)
+        let module_rel = if let Some(rest) = rel_str.strip_prefix("src/") {
+            rest
+        } else if let Some(rest) = rel_str.strip_prefix("test/") {
+            rest
+        } else if let Some(rest) = rel_str.strip_prefix("dev/") {
+            rest
+        } else {
+            return None;
+        };
+        let module_name = module_rel.strip_suffix(".gleam")?;
+
+        // A valid module filename is snake_case, starting with a lowercase letter
+        // and containing only [a-z0-9_] for each path segment
+        fn is_valid_module_basename(s: &str) -> bool {
+            let mut iter = s.chars();
+            match iter.next() {
+                Some(c) if c.is_ascii_lowercase() => {}
+                _ => return false,
+            }
+            iter.all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_')
+        }
+
+        let is_valid_full = module_name.split('/').all(is_valid_module_basename);
+        if is_valid_full {
+            return None;
+        }
+
+        // Compute a suggested valid name for each segment and join with '/'
+        let suggestion = module_name
+            .split('/')
+            .map(|seg| {
+                let mut s = to_snake_case(&ecow::EcoString::from(seg)).to_string();
+                if s.is_empty() {
+                    s = "module".into();
+                }
+                if !s.chars().next().map(|c| c.is_ascii_lowercase()).unwrap_or(false) {
+                    s = format!("m{}", s);
+                }
+                s
+            })
+            .collect::<Vec<_>>()
+            .join("/");
+
+        let mut feedback = Feedback::default();
+        // Highlight the first line to ensure the diagnostic is visible in editors
+        let first_line_end = src.find('\n').unwrap_or(src.len());
+        let span_end: u32 = first_line_end.max(1).try_into().unwrap_or(1);
+        let diagnostic = Diagnostic {
+            title: format!("{module_name} is not a valid name for a Gleam module"),
+            text: format!("You should switch to {suggestion} instead."),
```

Comment:
Include what the rules are for module names, and don't use the word "should", especially in times like this when we don't know what the best option is.